### PR TITLE
Cargo Exports (and a very basic economy system)

### PR DESCRIPTION
**Pull request by Core0verload**
This PR adds a new exports system to cargo. This time, it's 100% less goof.

All existing exports are ported to the new system, Full list: crates, syndie documents, seeds, max reliability designs, tech levels, manifests and plasma.

A lot of new exports are in too. You can export all the materials (in form of sheets, floor tiles, rods or even coins), fueltanks and watertanks, beer kegs and some organic products (monkey/cat/corgi/goliath/ligger hides). And even more exports are coming soon.

It also adds a very basic economy system. For example, space cash can now be exported. It will add cash's value to cargo credits score. Coin values are tied to the cost of materials they are made from. For example, an iron coin is worth 1cr, a plasma coin is worth 100cr, etc. Yes, in 26 century, gold standard lives again. It's the only way to make station-produced coins balanced and useable.

:cl: CoreOverload
rscadd: A lot of items are now exportable in cargo.
rscadd: Use Export Scanner to check any item's export value.
/:cl:

TODO:
- [x] Remove the `world <<` debug info.
- [x] Make selling proc recursive.
- [x] Add support for contraband and emagged exports.
- [x] Export scanner.
- [x] More exports.

In upcoming PRs:
- [ ] Sell cannon.
- [ ] Money Bag refactor.
- [ ] Even more exports.

### CODE REVIEW

**Comment by Bawhoppen** [comment]
interesting, but does this readd goof's jew-gun?

**Comment by Core0verload** [comment]
It will come in a separate PR. New system makes it easy to add, but I don't want to repeat Goof's mistake and try to do too many things at once.

**Comment by coiax** [comment]
If we're doing economy, might want to balance the slot machines at some point. I don't think their odds are the house wins, long term, if you put enough coins in them, you will get a profit from jackpot or 4-in-a-row.

**Comment by Shapsy** [comment]
You'll probably need to remove printing free money from botany's money trees and the service borgs if you want space cash to count for cargo points

**Comment by MMMiracles** [comment]
@Shapsy even better, make it so sending in too much credits in a short period causes inflation and ruins the value of paper money. Cause hyper-inflation and recreate post-WWI Germany where spaceman drag around lockers of worthless space cash.

**Comment by coiax** [comment]
I don't know how much money botany can output with money trees. You could fix service borgs spamming Space Cash by just increasing the time it takes to dispense and the energy cost.

I'm sure there's a way to make the prices "react" so if you only dump iron coins, their price will go down. Maybe have the prices permanent from shift to shift (with the sanity caps and the option of admin resets in case someone manages to flood the market with something).

**Comment by Core0verload** [comment]
This isn't an open market. No inflation, no market floods. All prices are bolted down to the floor because we sell to Nanotrasen only.

You may unlock some other sales option, like contraband sales that accept some more items, or Syndicate that pays more for plasma and buys NT intel. But most of the sales are going to NT anyway.

About borgs - AFAIK they dump out something like 20 credits per click. Yes, you can sell those, but it would be quicker to sell an unused crate or fuel tank for 500 credits.

**Comment by KorPhaeron** [comment]
Could just make borgs spit out monopoly money or something if it becomes an issue

**Comment by coiax** [comment]
Well, if all the sales are logged, we can see if there anything that's over or underpriced.

Also, is this touching the quite ugly export code, with the checking of all the cargo manifests?

**Comment by Core0verload** [comment]
All the manifests are now export datums. Every single old export is now an export datum, see OP. The datums are simple but flexible enough to allow that.

**Comment by duncathan** [comment]
What's the benefit of this system over goof's?

**Comment by coiax** [comment]
I thought the main problems with goofcargo was the bluespace selling gun and the INSTANT GUN PURCHASES. Oh yeah, and also selling empty medkits.

If you change the magical selling gun to teleport the items onto the loading conveyer belt, that could be a way of selling items quickly while still requiring a loading/unloading stage on the shuttle?

**Comment by RandomMarine** [comment]
@Shapsy @KorPhaeron 
IF money tree farms become an issue. This is an opportunity for inter-departmental cooperation.

**Comment by coiax** [comment]
Automated conveyor based money farms when? Although, this reminds me of a topic that I brought up when it was fashionable to do lots of stock trading.

You can get a lot of cargo points yes, but no one ever seems to want to spend them. We need more useful materials in cargo crates, I guess.

File: code/modules/cargo/exports/sheets.dm [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+// Monkey hide. Cheap.
+/datum/export/stack/skin/monkey
+	cost = 150
+	unit_name = "monkey hide"
+	export_types = list(/obj/item/stack/sheet/animalhide/monkey)
+
+// Human skin. Illegal
+/datum/export/stack/skin/human
+	cost = 2000
```
Comment (KorPhaeron) [inline code/modules/cargo/exports/sheets.dm]: This seems like a very, very easy thing to farm.

**Comment by bgobandit** [comment]
The issue with goofcargo was buying things and turning around and selling them for a lot more, FROM THE SAME OUTLET

**Comment by coiax** [comment]
Oh, you could also buy raw bananaium and refine it, and then sell the refined bars for profits. Because the raw stuff was cheaper than the refined stuff, even though it should be the other way around.

File: code/modules/cargo/exports/materials.dm [hunk 2]
```diff
@@ -0,0 +1,84 @@
+/datum/export/material
+	cost = 5 // Cost per MINERAL_MATERIAL_AMOUNT, which is 2000cm3 as of April 2016.
```
Comment (Iamgoofball) [inline code/modules/cargo/exports/materials.dm]: Just a thought, we should run math to generate this in the New()

File: code/modules/cargo/packs.dm [hunk 3]
```diff
@@ -664,11 +664,9 @@
 
 /datum/supply_pack/medical/iv_drip
 	name = "IV Drip Crate"
-	cost = 3000
-	access = access_cmo
+	cost = 1000
 	contains = list(/obj/machinery/iv_drip)
 	crate_name = "iv drip crate"
-	crate_type = /obj/structure/closet/crate/secure
```
Comment (Iamgoofball) [inline code/modules/cargo/packs.dm]: christ why were these even so secure, wtf were you thinking pappers

File: code/modules/mining/ores_coins.dm [hunk 4]
```diff
@@ -263,72 +266,78 @@
 	if(cmineral)
 		name = "[cmineral] coin"
 
+/obj/item/weapon/coin/examine(mob/user)
+	..()
+	if(value)
+		user << "<span class='info'>It's worth [value] credit\s.</span>"
```
Comment (Iamgoofball) [inline code/modules/mining/ores_coins.dm]: should add money goggles that let you see the current market value of shit when you examine them

**Comment by Iamgoofball** [comment]
Hnnngh

It's like my shit but better

:+1: x :100:

**Comment by PKPenguin321** [comment]
if buying unrefined, refining, and selling for higher is a problem, then just make unrefined cost more than refined and make it so you can only buy already refined materials. why would companies even sell unrefined minerals in the first place

**Comment by Iamgoofball** [comment]
mining companies

**Comment by Core0verload** [comment]
For now, buying and refining ore is not the problem, because the exports list is actually export only, no import for now. You can't just buy those items.

File: code/modules/mining/ores_coins.dm (hunk 4 above)
Comment (Core0verload) [inline code/modules/mining/ores_coins.dm]: Maybe in future PRs.

**Comment by Core0verload** [comment]
This is ready now.

**Comment by optimumtact** [comment]
please don't re add that silly gun that shot cashbeams and vanished shit on the spot.

at least try to make it slightly realistic

**Comment by Bawhoppen** [comment]
@optimumtact jew-gun best gun

**Comment by Core0verload** [comment]
Jew Gun will be re-added, but as emaggable crate, and it will be quite costly. But the gun's effect on a random redshirt is impressive:
Keira Stern is hit by an export ray in the chest!
+100 credits: Received 1 helmet.
+100 credits: Received 1 armor vest.
+5 credits: Received 1 radio.
+100 credits: Received 1 stun baton.
+250 credits: Received 1 advanced taser.
+10 credits: Received 1 handheld flash.
+3 credits: Received 1 pair of handcuffs.

**Comment by coiax** [comment]
The remote selling gun needs to be charged up though, because if it's just a projectile weapon, and you hit someone with it, then that's essentinally high range non-lethal polyacid. Although, that sounds like a great idea for a badmin version of the cargo gun

**Comment by Iamgoofball** [comment]
@Core0verload 
I like the idea but instead of being an emag only crate, it should be available as an EXPENSIVE crate and then emaggable for the "sell people's shit right off of their bodies" setup

**Comment by optimumtact** [comment]
I don't like it

**Comment by Bawhoppen** [comment]
yea okay but lets save the jew-gun for a seperate pr if you're going to be doing that shit

**Comment by Core0verload** [comment]
Yeah, this PR is ready. X-porter cannon will be added in another PR.

File: code/modules/cargo/exports/sheets.dm [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+	unit_name = ""
+
+// Monkey hide. Cheap.
+/datum/export/stack/skin/monkey
+	cost = 150
+	unit_name = "monkey hide"
+	export_types = list(/obj/item/stack/sheet/animalhide/monkey)
+
+// Human skin. Illegal
+/datum/export/stack/skin/human
```
Comment (KorPhaeron) [inline code/modules/cargo/exports/sheets.dm]: This seems very easy to gain points with

Comment (Core0verload) [inline code/modules/cargo/exports/sheets.dm]: It _seems_. You need a gibber (chief), some human bodies (dead players, or monkey cubes and help from geneticist/chemist) and a lot of patience. And what's your profit in the end? 1 human skin = 4 plasma sheets = 4 crates = 4 fuel tanks. Not that much.

**Comment by Core0verload** [comment]
Remove WIP tag already.

**Comment by coiax** [comment]
Why isn't this merged yet?

**Comment by Iamgoofball** [comment]
>Why isn't this merged yet?

File: code/modules/shuttle/supply.dm [hunk 6]
```diff
@@ -102,127 +109,42 @@
...
-	if(intel > 0)
-		pointsEarned = round(intel * SSshuttle.points_per_intel)
-		msg += "[pointsEarned]: Received [intel] article(s) of enemy intelligence."
-		SSshuttle.points += pointsEarned
+	for(var/a in exports)
+		var/datum/export/E = a
+		var/export_text = E.total_printout()
+		if(!export_text)
+			continue
+		world << export_text
```
Comment (AnturK) [inline code/modules/shuttle/supply.dm]: world <<

Comment (coiax) [inline code/modules/shuttle/supply.dm]: Everyone gets to know about the exporting! It's a feature!

**Comment by Core0verload** [comment]
I realized that all the time I worked on another branch. Now I pushed the changes and it's actually ready.

**Pull request merged**